ibm/resource_ibm_is_snapshot.go

```diff
@@ -0,0 +1,483 @@
+
+	validateSchema := make([]ValidateSchema, 1)
+	validateSchema = append(validateSchema,
+		ValidateSchema{
+			Identifier:                 isSnapshotName,
+			ValidateFunctionIdentifier: ValidateRegexpLen,
+			Type:                       TypeString,
+			Required:                   true,
+			Regexp:                     `^([a-z]|[a-z][-a-z0-9]*[a-z0-9])$`,
+			MinValueLength:             1,
+			MaxValueLength:             63})
+	ibmISSnapshotResourceValidator := ResourceValidator{ResourceName: "ibm_is_snapshot", Schema: validateSchema}
+	return &ibmISSnapshotResourceValidator
+}
+
+func resourceIBMISSnapshotCreate(d *schema.ResourceData, meta interface{}) error {
+	sv := ""
+	rg := ""
+	name := ""
+	if sourceVolume, ok := d.GetOk(isSnapshotSourceVolume); ok {
+		sv = sourceVolume.(string)
+	}
+	if grp, ok := d.GetOk(isVPCResourceGroup); ok {
+		rg = grp.(string)
+	}
+	log.Printf("[DEBUG] Snapshot create")
+	if snapshotName, ok := d.GetOk(isSnapshotName); ok {
+		name = snapshotName.(string)
+	}
+	err := snapshotCreate(d, meta, name, sv, rg)
+	if err != nil {
+		return err
+	}
+	return resourceIBMISSnapshotRead(d, meta)
+}
+
+func snapshotCreate(d *schema.ResourceData, meta interface{}, name, sv, rg string) error {
+	sess, err := vpcClient(meta)
+	if err != nil {
+		return err
+	}
+
+	options := &vpcv1.CreateSnapshotOptions{
+		Name: &name,
+	}
+	options.SourceVolume = &vpcv1.VolumeIdentity{
+		ID: &sv,
+	}
+	if rg != "" {
+		options.ResourceGroup = &vpcv1.ResourceGroupIdentity{
+			ID: &rg,
+		}
+	}
+
+	snapshot, response, err := sess.CreateSnapshot(options)
+	if err != nil {
+		return fmt.Errorf("Error creating Snapshot %s\n%s", err, response)
+	}
+	d.SetId(*snapshot.ID)
+	log.Printf("[INFO] Snapshot : %s", *snapshot.ID)
+	_, err = isWaitForSnapshotAvailable(sess, d.Id(), d.Timeout(schema.TimeoutCreate))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func isWaitForSnapshotAvailable(sess *vpcv1.VpcV1, id string, timeout time.Duration) (interface{}, error) {
+	log.Printf("Waiting for Snapshot (%s) to be available.", id)
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{isSnapshotPending},
+		Target:     []string{isSnapshotAvailable, isSnapshotFailed},
+		Refresh:    isSnapshotRefreshFunc(sess, id),
+		Timeout:    timeout,
+		Delay:      10 * time.Second,
+		MinTimeout: 10 * time.Second,
+	}
+
+	return stateConf.WaitForState()
+}
+
+func isSnapshotRefreshFunc(sess *vpcv1.VpcV1, id string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		getSnapshotOptions := &vpcv1.GetSnapshotOptions{
+			ID: &id,
+		}
+		snapshot, response, err := sess.GetSnapshot(getSnapshotOptions)
+		if err != nil {
+			return nil, isSnapshotFailed, fmt.Errorf("Error getting Snapshot : %s\n%s", err, response)
+		}
+
+		if *snapshot.LifecycleState == isSnapshotAvailable || *snapshot.LifecycleState == isSnapshotFailed {
+			return snapshot, *snapshot.LifecycleState, nil
+		}
```

Comment:
Seperated support *snapshot.LifecycleState == isSnapshotFailed and return error  (snapshot, *snapshot.LifecycleState, fmt.Error("XXXXX")